pkg/ruler/rules/objectclient/rule_store.go

```diff
@@ -172,6 +163,58 @@ func (o *RuleStore) DeleteNamespace(ctx context.Context, userID, namespace strin
 	return nil
 }
 
+func (o *RuleStore) loadRuleGroupsConcurrently(ctx context.Context, rgObjects []chunk.StorageObject) (map[string]rules.RuleGroupList, error) {
+	ch := make(chan string, len(rgObjects))
+
+	for _, obj := range rgObjects {
+		level.Debug(util.Logger).Log("msg", "listing rule group", "key", obj.Key)
+		ch <- obj.Key
+	}
+	close(ch)
+
+	mtx := sync.Mutex{}
+	result := map[string]rules.RuleGroupList{}
+
+	concurrency := loadRuleGroupsConcurrency
+	if loadRuleGroupsConcurrency < len(rgObjects) {
+		concurrency = len(rgObjects)
+	}
+
+	// Given we store one file per rule group. With this, we create a pool of workers that will
+	// download all rule groups in parallel. We limit the number of workers to avoid a
+	// particular user having too many rule groups rate limiting us with the object storage.
+	g, gCtx := errgroup.WithContext(ctx)
+	for i := 0; i < concurrency; i++ {
+		g.Go(func() error {
+			for key := range ch {
+				user := decomposeRuleObjectKey(key)
+				if user == "" {
+					continue
+				}
+
+				rg, err := o.getRuleGroup(gCtx, key)
+				if err != nil {
+					level.Error(util.Logger).Log("msg", "failed to get rule group", "key", key, "user", user)
+					return err
+				}
+
+				mtx.Lock()
+				result[user] = append(result[user], rg)
+				mtx.Unlock()
+			}
+
+			return nil
+		})
+	}
+
+	err := g.Wait()
+	if err != nil {
+		return map[string]rules.RuleGroupList{}, err
+	}
+
+	return result, nil
```

Comment:
```suggestion
	return result, err
```

Reply:
Doing it like this would return partial results on `result`, but I guess it does not really matter given uses of this check for  `err` and return immediately.